core/src/idx/trees/store/mapper.rs

```diff
@@ -0,0 +1,75 @@
+use crate::err::Error;
+use crate::idx::ft::analyzer::mapper::Mapper;
+use crate::sql::statements::DefineAnalyzerStatement;
+use crate::sql::Filter;
+use ahash::HashSet;
+use dashmap::DashMap;
+use std::path::Path;
+
+#[derive(Default)]
+pub(crate) struct Mappers(DashMap<String, Mapper>);
+
+impl Mappers {
+	/// If any mapper is defined, it will be loaded in memory.
+	pub(crate) async fn load(&self, az: &DefineAnalyzerStatement) -> Result<(), Error> {
+		if let Some(filters) = &az.filters {
+			for f in filters {
+				if let Filter::Mapper(path) = f {
+					self.insert(path).await?;
+				}
+			}
+		}
+		Ok(())
+	}
+
+	/// Ensure that if a mapper is defined, that it is also loaded in memory.
+	/// This method does not reload a mapper if it is already in memory.
+	pub(crate) async fn check(&self, az: &DefineAnalyzerStatement) -> Result<(), Error> {
+		if let Some(filters) = &az.filters {
+			for f in filters {
+				if let Filter::Mapper(path) = f {
+					if !self.0.contains_key(path) {
+						self.insert(path).await?;
+					}
+				}
+			}
+		}
+		Ok(())
+	}
+
+	async fn insert(&self, path: &str) -> Result<(), Error> {
+		let p = Path::new(path);
+		if !p.exists() || !p.is_file() {
+			return Err(Error::Internal(format!("Invalid mapper path: {p:?}")));
+		}
```

Comment:
The file is immediately opened and read in the line following this one.
(not sure I get the point right :D)

https://github.com/surrealdb/surrealdb/blob/98b694c6093deb134dc69e67f7f219aae9d901ce/core/src/idx/trees/store/mapper.rs#L45

Reply:
I know it is maybe a bit too pedantic, but you can get time of check time of use issues with first checking existence and then opening. It is possible for the file to change between the call to exists() and open in which case a different error would happen. The thread can be swapped out, or tokio might decide that the task has run for to long and make the file::open future return pending to swap out the task. In the mean time the file might change. 

It is probably not a big deal since the only thing that will happen is that you get a IoError instead of Internal error.